torch/csrc/Stream.cpp

```diff
@@ -256,6 +256,73 @@ static PyObject* THPStream_eq(THPStream* self, THPStream* other) {
   END_HANDLE_TH_ERRORS
 }
 
+static PyObject* THPStream_enter(THPStream* self, PyObject* unused) {
+  HANDLE_TH_ERRORS
+  // No operation is performed if the stream does not belong to an accelerator.
+  c10::DeviceType dst_device_type =
+      static_cast<c10::DeviceType>(self->device_type);
+  if (C10_UNLIKELY(!at::accelerator::isAccelerator(dst_device_type))) {
```

Comment:
I'm not sure why this limitation exists? I would expect the logic here to be equivalent to:
```python
new_stream = ...
try:
  old_stream = get_current_stream(new_stream.device_type, new_stream.device_idx) # this API doesn't exist, but just to show
  set_current_stream(new_stream.device_type, new_stream.device_idx, new_stream)
  # User code
finally:
    set_current_stream(new_stream.device_type, new_stream.device_idx, old_stream)


```

Reply:
Because `new_stream` might contain the device index information that is not equal to the current device index, we have to switch the device index to `new_stream.device_index` when `try` and change it back when `finally`.
I simplify the logic like this:
```python
new_stream = ...
try:
    **src_device** = torch.accelerator.current_device_idx()
    dst_device = new_stream.device_index
    # switch the current device index to dst_device
    torch.accelerator.set_device_idx(dst_device)
    # fetch the previously current stream on dst_device
    **dst_prev_stream** = torch.accelerator.current_stream()
    # set the current stream to new_stream
    torch.accelerator.set_stream(new_stream)
    # user code
finally:
    # switch current stream on target device back to original
    torch.accelerator.set_stream(**dst_prev_stream**)
    # switch current device to original device
    torch.accelerator.set_device(**src_device**)
```

Reply:
Is the crux of the difference between the two things here that in your case, doing "with stream" changes the current device idx while my proposal above was only changing the current stream for the device idx associated with that stream without changing the current device idx.

Reply:
I concern this will raise some runtime error and unexpected behavior for CUDA if not changing the current device idx. For a simple example, 
```python
with stream:
    # user code
    torch.cuda.synchronize()
```
This will make `torch.cuda.synchronize` to sync the current device idx instead of device idx inside stream. 

Furthermore, in my understanding, when using `cudaLaunchKernel`, we need to ensure the current device matches the device associated with the stream. (refer to https://stackoverflow.com/questions/37490869/does-cudasetdevice-need-to-be-called-prior-to-kernel-launch-with-stream) This is because in CUDA, devices and streams are tightly coupled, and kernel execution must occur within the device context of the specified stream. Essentially, switching the current device idx to stream device idx help avoid runtime error due to the inconsistent runtime context.
If `with stream` not change the current device, all CUDA kernels in pytorch must call `c10::cuda::SetDevice` before the kernel launch. Meanwhile, I am afraid this will impact other scenario like, CUDAEvent and ipc, etc.